File: L17_Find_the_starting_point_of_the_Loop_or_Cycle_in_LinkedList.java
class Node {
    public int data;
    public Node next;

    Node()
    {
        this.data = 0;
        this.next = null;
    }
    Node(int data)
    {
        this.data = data;
        this.next = null;
    }
    Node(int data, Node next)
    {
        this.data = data;
        this.next = next;
    }
}
public class L17_Find_the_starting_point_of_the_Loop_or_Cycle_in_LinkedList {
    public static Node firstNode(Node head) {
        // Write your code here.
        Node slow = head;
        Node fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast){
                slow = head;
                while(slow != fast){
                    slow = slow.next;
                    fast = fast.next;
                }return slow;
            }
        }
        return null;
    }
}

File: L17_Find_the_starting_point_of_the_Loop_or_Cycle_in_LinkedList.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

class Node{
public:
    int data;
    Node *next;
    Node()
    {
        this->data = 0;
        next = NULL;
    }
    Node(int data)
    {
        this->data = data;
        this->next = NULL;
    }
    Node(int data, Node* next)
    {
        this->data = data;
        this->next = next;
    }
};

Node *firstNode(Node *head){
    //    Write your code here.
    Node *slow = head, *fast = head;
    while(fast && fast->next){
        slow = slow->next;
        fast = fast->next->next;
        if(slow == fast){
            slow = head;
            while(slow != fast){
                slow = slow->next;
                fast = fast->next;
                //if(slow == fast) 
            }return slow;
        }
    }
    return nullptr;
}